### PR TITLE
logging: frontends: stmesp: Add turbo logs

### PR DESCRIPTION
**Pull request by nordic-krch**
Add 'turbo' logging feature. When enabled, short logs (no argument or one numeric, 32 bit argument) are handled in a special way that is much faster than the default one (>10x faster). Additionally, there is an option to remove all other logs from the system which allows to not include almost any logging framework code in the binary (~170 bytes of code is needed). It may be especially valuable for memory constraint targets (ppr, flpr) where with only 170 byte of code (+code for each log message) we can provide limited formatted string logging support.

 'Turbo' logging is using following tricks to achieve that:
- logging strings are put into a memory section and additional memory section is created which holds addresses of those strings. Index in that array is used to identify a string (32 bit address is encoded into a smaller number, 15 bits is more than enough). This index is used for a STMESP register set (there are 2^16 available). STMESP channel encodes string.
- Logging level is stringified and prepended to a string
- Source ID is encoded by using DM16 (so far not used).
- Log without arguments is written as DMTS16
- Log with one argumetn is written as DM16+DMTS32

`samples/boards/nordic/coresight_stm`
Without turbo logs:
```
flpr: Timing for log message with 0 arguments: 1.20us
flpr: Timing for log message with 1 argument: 2.20us
flpr: Timing for log message with 2 arguments: 1.80us
flpr: Timing for log message with 3 arguments: 2.30us
flpr: Timing for log_message with string: 3.80us
flpr: Timing for tracepoint: 0.10us
flpr: Timing for tracepoint_d32: 0.10us
app: Timing for log message with 0 arguments: 1.80us
app: Timing for log message with 1 argument: 1.90us
app: Timing for log message with 2 arguments: 2.0us
app: Timing for log message with 3 arguments: 2.10us
app: Timing for log_message with string: 4.40us
app: Timing for tracepoint: 0.5us
app: Timing for tracepoint_d32: 0.10us
ppr: Timing for log message with 0 arguments: 25.40us
ppr: Timing for log message with 1 argument: 27.20us
ppr: Timing for log message with 2 arguments: 27.70us
ppr: Timing for log message with 3 arguments: 27.60us
ppr: Timing for log_message with string: 65.50us
ppr: Timing for tracepoint: 0.65us
ppr: Timing for tracepoint_d32: 0.30us
```

and with turbo logs:
```
flpr: Timing for log message with 0 arguments: 0.10us
flpr: Timing for log message with 1 argument: 0.20us
flpr: Timing for log message with 2 arguments: 2.20us
flpr: Timing for log message with 3 arguments: 2.20us
flpr: Timing for log_message with string: 3.90us
flpr: Timing for tracepoint: 0.0us
flpr: Timing for tracepoint_d32: 0.10us
app: Timing for log message with 0 arguments: 0.10us
app: Timing for log message with 1 argument: 0.20us
app: Timing for log message with 2 arguments: 2.10us
app: Timing for log message with 3 arguments: 2.0us
app: Timing for log_message with string: 4.60us
app: Timing for tracepoint: 0.5us
app: Timing for tracepoint_d32: 0.5us
ppr: Timing for log message with 0 arguments: 2.10us
ppr: Timing for log message with 1 argument: 3.10us
ppr: Timing for log message with 2 arguments: 27.80us
ppr: Timing for log message with 3 arguments: 27.40us
ppr: Timing for log_message with string: 68.20us
ppr: Timing for tracepoint: 0.55us
ppr: Timing for tracepoint_d32: 0.30us
```

~~It is based on #79419 and #79420.~~

### CODE REVIEW

**Comment by nordic-krch** [comment]
@nordic-segl fyi. If that goes in then test for coresight_stm will need to be updated with new timings.

**Comment by nordic-krch** [comment]
@kl-cruz @masz-nordic can you take a look?

File: include/zephyr/logging/log_frontend_stmesp_demux.h [hunk 1]
```diff
@@ -121,11 +121,14 @@ struct log_frontend_stmesp_demux_trace_point {
 	/** Flag indicating if trace point includes data. */
 	uint64_t has_data: 1;
 
-	/** Timestamp. */
-	uint64_t timestamp: 58;
+	/** Timestamp. 54 bits at 40MHz is >14 years. */
+	uint64_t timestamp: 54;
```
Comment (kl-cruz) [inline include/zephyr/logging/log_frontend_stmesp_demux.h:125]: I guess bitfield operation, especially non-aligned to 2^x can slow down operations a bit. How often is this structure used?

Comment (nordic-krch) [reply include/zephyr/logging/log_frontend_stmesp_demux.h:125]: It's not used much. Data is written when decoding STPv2 stream (STM data) and then read to decode it to human readable format.

File: subsys/logging/frontends/Kconfig [hunk 2]
```diff
@@ -67,6 +67,28 @@ config LOG_FRONTEND_STMESP_EARLY_BUF_SIZE
 config LOG_FRONTEND_STMESP_DICT_VER
 	def_int 0
 
+config LOG_FRONTEND_STMESP_TURBO_LOG
+	bool "Optimize short_logs"
+	select LOG_CUSTOM_HEADER
+	default y if (NRF_ETR || SOC_NRF54H20_CPUPPR || SOC_NRF54H20_CPUFLPR) \
```
Comment (masz-nordic) [inline subsys/logging/frontends/Kconfig:73]: Lumos FLPRs as well?

**Review by masz-nordic — APPROVED** [review]

File: subsys/logging/frontends/Kconfig (hunk 2 above)
Comment (nordic-krch) [reply subsys/logging/frontends/Kconfig:73]: no, Lumos does not have STM.

**Review by kl-cruz — APPROVED** [review]

**Comment by kartben** [comment]
@nordic-krch Can you please look into failure in main

https://github.com/zephyrproject-rtos/zephyr/actions/runs/12253888590/job/34183618675#step:11:1018

```
[31/178] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/debug/coresight/cs_trace_defmt.c.obj
[32/178] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/ipc/ipc_service/ipc_service.c.obj
FAILED: zephyr/CMakeFiles/zephyr.dir/subsys/ipc/ipc_service/ipc_service.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DNRF54H20_XXAA -DNRF_APPLICATION -DPICOLIBC_DOUBLE_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/kernel/include -I/__w/zephyr/zephyr/arch/arm/include -I/__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/coresight_stm/zephyr/include/generated/zephyr -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/coresight_stm/zephyr/include/generated -I/__w/zephyr/zephyr/soc/nordic -I/__w/zephyr/zephyr/soc/nordic/nrf54h/. -I/__w/zephyr/zephyr/soc/nordic/nrf54h/gpd/include -I/__w/zephyr/zephyr/soc/nordic/common/. -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/zephyr/modules/cmsis/. -I/__w/zephyr/modules/hal/nordic/nrfx -I/__w/zephyr/modules/hal/nordic/nrfx/drivers/include -I/__w/zephyr/modules/hal/nordic/nrfx/mdk -I/__w/zephyr/zephyr/modules/hal_nordic/nrfx/. -I/__w/zephyr/modules/hal/nordic/nrfs/include -I/__w/zephyr/modules/hal/nordic/nrfs/include/services -I/__w/zephyr/modules/hal/nordic/nrfs/helpers -I/__w/zephyr/zephyr/modules/hal_nordic/nrfs/. -I/__w/zephyr/zephyr/modules/hal_nordic/nrfs/backends -I/__w/zephyr/zephyr/modules/hal_nordic/nrfs/dvfs -isystem /__w/zephyr/zephyr/lib/libc/common/include -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/coresight_stm/zephyr/include/generated/zephyr/autoconf.h -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m33 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/boards/nordic/coresight_stm=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=picolibc.specs -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/subsys/ipc/ipc_service/ipc_service.c.obj -MF zephyr/CMakeFiles/zephyr.dir/subsys/ipc/ipc_service/ipc_service.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/subsys/ipc/ipc_service/ipc_service.c.obj -c /__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c
In file included from /__w/zephyr/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /__w/zephyr/zephyr/include/zephyr/irq_multilevel.h:16,
                 from /__w/zephyr/zephyr/include/zephyr/devicetree.h:20,
                 from /__w/zephyr/zephyr/include/zephyr/device.h:12,
                 from /__w/zephyr/zephyr/include/zephyr/ipc/ipc_service.h:11,
                 from /__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c:8:
/__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c: In function 'ipc_service_register_endpoint':
/__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c:83:63: error: passing argument 3 of 'log_frontend_stmesp_log1' makes integer from pointer without a cast [-Werror=int-conversion]
   83 |         LOG_DBG("Register endpoint %s", cfg->name ? cfg->name : "");
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:61:9: note: in expansion of macro '__COND_CODE'
   61 |         __COND_CODE(_ZZZZ##_flag, _if_0_code, _else_code)
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_macro.h:212:9: note: in expansion of macro 'Z_COND_CODE_0'
  212 |         Z_COND_CODE_0(_flag, _if_0_code, _else_code)
      |         ^~~~~~~~~~~~~
/__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:90:9: note: in expansion of macro 'COND_CODE_0'
   90 |         COND_CODE_0(NUM_VA_ARGS_LESS_1(__VA_ARGS__),                                     \
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:69:53: note: in expansion of macro '__DEBRACKET'
   69 | #define __GET_ARG2_DEBRACKET(ignore_this, val, ...) __DEBRACKET val
      |                                                     ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_macro.h:196:9: note: in expansion of macro 'Z_COND_CODE_1'
  196 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:92:17: note: in expansion of macro 'COND_CODE_1'
   92 |                 COND_CODE_1(NUM_VA_ARGS_LESS_1(__VA_ARGS__),                             \
      |                 ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/sys/util_macro.h:371:27: note: in expansion of macro 'Z_GET_ARG_2'
  371 | #define GET_ARG_N(N, ...) Z_GET_ARG_##N(__VA_ARGS__)
      |                           ^~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/logging/log_frontend_stmesp.h:135:67: note: in expansion of macro 'GET_ARG_N'
  135 |                 log_frontend_stmesp_log1(_source, idx, (uintptr_t)GET_ARG_N(2, __VA_ARGS__));      \
      |                                                                   ^~~~~~~~~
/__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:77:17: note: in expansion of macro 'LOG_FRONTEND_STMESP_LOG1'
   77 |                 LOG_FRONTEND_STMESP_LOG1(_source, STRINGIFY(_level) __VA_ARGS__, dummy);           \
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:93:26: note: in expansion of macro 'Z_LOG_STMESP_1'
   93 |                         (Z_LOG_STMESP_1(_level, _source, __VA_ARGS__)),                  \
      |                          ^~~~~~~~~~~~~~
/__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:103:22: note: in expansion of macro 'Z_LOG_STMESP'
  103 | #define LOG_DBG(...) Z_LOG_STMESP(LOG_LEVEL_DBG, Z_LOG_CURRENT_DATA(), __VA_ARGS__)
      |                      ^~~~~~~~~~~~
/__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c:83:9: note: in expansion of macro 'LOG_DBG'
   83 |         LOG_DBG("Register endpoint %s", cfg->name ? cfg->name : "");
      |         ^~~~~~~
In file included from /__w/zephyr/zephyr/soc/nordic/common/./zephyr_custom_log.h:14,
                 from /__w/zephyr/zephyr/include/zephyr/logging/log.h:458,
                 from /__w/zephyr/zephyr/subsys/ipc/ipc_service/ipc_service.c:11:
/__w/zephyr/zephyr/include/zephyr/logging/log_frontend_stmesp.h:100:72: note: expected 'uint32_t' {aka 'unsigned int'} but argument is of type 'const char *'
  100 | void log_frontend_stmesp_log1(const void *source, uint32_t x, uint32_t arg);
      |                                                               ~~~~~~~~~^~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
FAILED: _sysbuild/sysbuild/images/coresight_stm-prefix/src/coresight_stm-stamp/coresight_stm-build /__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/_sysbuild/sysbuild/images/coresight_stm-prefix/src/coresight_stm-stamp/coresight_stm-build 
cd /__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/coresight_stm && /usr/bin/cmake --build .
ninja: build stopped: subcommand failed.

INFO    - /__w/zephyr/zephyr/twister-out/nrf54h20dk@0.9.0_nrf54h20_cpuapp/samples/boards/nordic/coresight_stm/sample.boards.nrf.coresight_stm/build.log
INFO    -  235/1117 nrf54h20dk@0.9.0/nrf54h20/cpuapp drivers.flash.common.no_explicit_erase             NOT RUN (build)
INFO    -  236/1117 nrf54h20dk@0.9.0/nrf54h20/cpuapp drivers.comparator.gpio_loopback.nrf_lpcomp        NOT RUN (build)
INFO    -  237/1117 nrf54h20dk@0.9.0/nrf54h20/cpuapp drivers.comparator.gpio_loopback.nrf_comp          NOT RUN (build)
INFO    -  238/1117 nrf54h20dk@0.9.0/nrf54h20/cpuapp dri
```

**Comment by kartben** [comment]
@nordic-krch another one:

```
logging.log_custom_header on qemu_cortex_m0/nrf51822 error (Build failure)

[9/149] Building C object zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj
[10/149] Generating include/generated/zephyr/offsets.h
[11/149] Building C object CMakeFiles/app.dir/src/log_custom_header.c.obj
FAILED: CMakeFiles/app.dir/src/log_custom_header.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DNRF51 -DNRF51422_XXAA -DPICOLIBC_LONG_LONG_PRINTF_SCANF -DTC_RUNID=7ba3cfd1b3be43cfd7666305c640e8a4 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/twister-out/qemu_cortex_m0_nrf51822/tests/subsys/logging/log_custom_header/logging.log_custom_header/zephyr/include/generated/zephyr -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/qemu_cortex_m0_nrf51822/tests/subsys/logging/log_custom_header/logging.log_custom_header/zephyr/include/generated -I/__w/zephyr/zephyr/soc/nordic -I/__w/zephyr/zephyr/soc/nordic/nrf51/. -I/__w/zephyr/zephyr/soc/nordic/common/. -I/__w/zephyr/zephyr/subsys/testsuite/include -I/__w/zephyr/zephyr/subsys/testsuite/coverage -I/__w/zephyr/zephyr/subsys/testsuite/ztest/include -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/zephyr/modules/cmsis/. -I/__w/zephyr/modules/hal/nordic/nrfx -I/__w/zephyr/modules/hal/nordic/nrfx/drivers/include -I/__w/zephyr/modules/hal/nordic/nrfx/mdk -I/__w/zephyr/zephyr/modules/hal_nordic/nrfx/. -I/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src -isystem /__w/zephyr/zephyr/lib/libc/common/include -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/qemu_cortex_m0_nrf51822/tests/subsys/logging/log_custom_header/logging.log_custom_header/zephyr/include/generated/zephyr/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m0 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=picolibc.specs -std=c99 -MD -MT CMakeFiles/app.dir/src/log_custom_header.c.obj -MF CMakeFiles/app.dir/src/log_custom_header.c.obj.d -o CMakeFiles/app.dir/src/log_custom_header.c.obj -c /__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c
In file included from /__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest.h:42,
                 from /__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:14:
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c: In function 'log_custom_header_test_macro_prefix':
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:58:35: error: 'CUSTOM_LOG_PREFIX' undeclared (first use in this function)
   58 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "DBG 0", strlen(output));
      |                                   ^~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:180:35: note: in definition of macro '_zassert_base'
  180 |                         z_zassert(cond, _msg ? ("(" default_msg ")") : (default_msg), __FILE__,    \
      |                                   ^~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:194:9: note: in expansion of macro '_zassert_va'
  194 |         _zassert_va(cond, default_msg, COND_CODE_1(__VA_OPT__(1), (__VA_ARGS__), (NULL)))
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:9: note: in expansion of macro 'zassert'
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |         ^~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:382:32: note: in expansion of macro 'zassert_mem_equal__'
  382 | #define zassert_mem_equal(...) zassert_mem_equal__(__VA_ARGS__)
      |                                ^~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:58:9: note: in expansion of macro 'zassert_mem_equal'
   58 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "DBG 0", strlen(output));
      |         ^~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:58:35: note: each undeclared identifier is reported only once for each function it appears in
   58 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "DBG 0", strlen(output));
      |                                   ^~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:180:35: note: in definition of macro '_zassert_base'
  180 |                         z_zassert(cond, _msg ? ("(" default_msg ")") : (default_msg), __FILE__,    \
      |                                   ^~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:194:9: note: in expansion of macro '_zassert_va'
  194 |         _zassert_va(cond, default_msg, COND_CODE_1(__VA_OPT__(1), (__VA_ARGS__), (NULL)))
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:9: note: in expansion of macro 'zassert'
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |         ^~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:382:32: note: in expansion of macro 'zassert_mem_equal__'
  382 | #define zassert_mem_equal(...) zassert_mem_equal__(__VA_ARGS__)
      |                                ^~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:58:9: note: in expansion of macro 'zassert_mem_equal'
   58 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "DBG 0", strlen(output));
      |         ^~~~~~~~~~~~~~~~~
   73 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "ERR 1 2 3", strlen(output));
      |         ^~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:23: note: to match this '('
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |                       ^
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:180:35: note: in definition of macro '_zassert_base'
  180 |                         z_zassert(cond, _msg ? ("(" default_msg ")") : (default_msg), __FILE__,    \
      |                                   ^~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:194:9: note: in expansion of macro '_zassert_va'
  194 |         _zassert_va(cond, default_msg, COND_CODE_1(__VA_OPT__(1), (__VA_ARGS__), (NULL)))
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:9: note: in expansion of macro 'zassert'
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |         ^~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:382:32: note: in expansion of macro 'zassert_mem_equal__'
  382 | #define zassert_mem_equal(...) zassert_mem_equal__(__VA_ARGS__)
      |                                ^~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:73:9: note: in expansion of macro 'zassert_mem_equal'
   73 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "ERR 1 2 3", strlen(output));
      |         ^~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:17: error: too few arguments to function 'memcmp'
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |                 ^~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:180:35: note: in definition of macro '_zassert_base'
  180 |                         z_zassert(cond, _msg ? ("(" default_msg ")") : (default_msg), __FILE__,    \
      |                                   ^~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:194:9: note: in expansion of macro '_zassert_va'
  194 |         _zassert_va(cond, default_msg, COND_CODE_1(__VA_OPT__(1), (__VA_ARGS__), (NULL)))
      |         ^~~~~~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:396:9: note: in expansion of macro 'zassert'
  396 |         zassert(memcmp(buf, exp, size) == 0, #buf " not equal to " #exp, ##__VA_ARGS__)
      |         ^~~~~~~
/__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:382:32: note: in expansion of macro 'zassert_mem_equal__'
  382 | #define zassert_mem_equal(...) zassert_mem_equal__(__VA_ARGS__)
      |                                ^~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/tests/subsys/logging/log_custom_header/src/log_custom_header.c:73:9: note: in expansion of macro 'zassert_mem_equal'
   73 |         zassert_mem_equal(output, CUSTOM_LOG_PREFIX "ERR 1 2 3", strlen(output));
      |         ^~~~~~~~~~~~~~~~~
/opt/toolchains/zephyr-sdk-0.17.0/arm-zephyr-eabi/picolibc/include/string.h:84:10: note: declared here
   84 | int      memcmp (const void *, const void *, size_t);
      |          ^~~~~~
ninja: build stopped: subcommand failed.

INFO    - /__w/zephyr/zephyr/twister-out/qemu_cortex_m0_nrf51822/tests/subsys/logging/log_custom_header/logging.log_custom_header/build.log
INFO    -   61/1117 qemu_cortex_m0/nrf51822   logging.output.custom_timestamp                    PASSED (qemu 1.168s)
INFO    -   62/1117 qemu_cortex_m0/nrf51822   logging.output                                     PASSED (qemu 1.224s)
INFO    -   63/1117 qemu_cortex_m0/nrf51822   logging.mipi_syst.text                             PASSED (qemu 1.040s)
INFO    -   64/1117 qemu_cortex_m0/nrf51822   logging.mipi_syst                                  PASSED (qemu 1.074s)


```

**Comment by nordic-krch** [comment]
Second issue was caused by putting `zephyr_custom_log.h` header which is specific for that type of logging in a path which was added to every nordic target. It lead to compile error when compiling a test which was using "custom logging header" feature and was expecting to use custom header created for the test but instead the one added by this PR was used.

Fix #82795.

**Pull request merged**